### PR TITLE
Implement image export

### PR DESCRIPTION
**Pull request by weeman1337**
How a developer test could look like

- Create a NeoBoard with an image
- Export the board
- Extract and decode the image, e.g. `cat NeoBoard.nwb | jq -r ".whiteboard.files[0].data" | base64 --decode > image`
- Open the file with an image viewer
- It should show the image

**:heavy_check_mark: Checklist**

- [x] A changeset describing the change and affected packages ([more info](https://github.com/nordeck/.github/blob/main/docs/CONTRIBUTING.md#changelog-and-versioning)).
- [ ] Added or updated documentation.
- [x] Tests for new functionality and regression tests for bug fixes.
- [ ] Screenshots or videos attached (for UI changes).
- [x] All your commits have a `Signed-off-by` line in the message ([more info](https://github.com/nordeck/.github/blob/main/docs/CONTRIBUTING.md#dco)).

### CODE REVIEW

File: src/state/export/exportWhiteboard.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+      })),
+      files: await exportFiles(doc, baseUrl),
+    },
+  };
+}
+
+async function exportFiles(
+  doc: SharedMap<WhiteboardDocument>,
+  baseUrl: string,
+): Promise<WhiteboardFileExport[]> {
```
Comment (maheichyk) [inline src/state/export/exportWhiteboard.ts:65]: a suggestion: to return `undefined` if no images are exported

File: src/components/BoardBar/ExportWhiteboardDialogDownloadFile.tsx [hunk 2]
```diff
@@ -14,54 +14,56 @@
...
 
-    return () => {
-      URL.revokeObjectURL(url);
-    };
-  }, [whiteboardInstance]);
+  const handleDownloadClick = useCallback(async () => {
+    setIsDownloading(true);
+    const filename = `${roomName}.nwb`;
+    const whiteboardData = await whiteboard.export(
+      widgetApi.widgetParameters.baseUrl ?? '',
```
Comment (maheichyk) [inline src/components/BoardBar/ExportWhiteboardDialogDownloadFile.tsx:54]: `widgetApi.widgetParameters.baseUrl` have to be already defined there, better to throw an error as we do when image displays.

File: src/state/export/exportWhiteboard.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  return {
+    version: 'net.nordeck.whiteboard@v1',
+    whiteboard: {
+      slides: getNormalizedSlideIds(doc).map((slideId) => ({
+        elements: getNormalizedElementIds(doc, slideId)
+          .map((elementId) => getElement(doc, slideId, elementId)?.toJSON())
+          .filter(isDefined),
+        lock: getSlideLock(doc, slideId) ? {} : undefined,
+      })),
+      files: await exportFiles(doc, baseUrl),
```
Comment (maheichyk) [inline src/state/export/exportWhiteboard.ts:57]: Should we care that if image id duplicated we could have a one entry in `files`?

**Review by maheichyk — CHANGES_REQUESTED** [review]
Looks good to me 👍, few smaller points to discuss

File: src/state/export/exportWhiteboard.ts (hunk 3 above)
Comment (weeman1337) [reply src/state/export/exportWhiteboard.ts:57]: Good point. There should be a deduplication in the export data.

File: .changeset/orange-lions-invent.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@nordeck/matrix-neoboard-widget': minor
+---
+
+Export board with uoloaded images
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply .changeset/orange-lions-invent.md:5]: :facepalm:

File: yarn.lock [hunk 5]
```diff
@@ -292,11 +399,26 @@
     chalk "^2.4.2"
     js-tokens "^4.0.0"
 
+"@babel/highlight@^7.24.2":
+  version "7.24.2"
+  resolved "https://registry.yarnpkg.com/@babel/highlight/-/highlight-7.24.2.tgz#3f539503efc83d3c59080a10e6634306e0370d26"
+  integrity sha512-Yac1ao4flkTxTteCDZLEvdxg2fZfz1v8M4QpaGypq/WPDqg3ijHYbDfs+LG5hvzSoqaSZ9/Z9lKSP3CjZjv+pA==
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply yarn.lock:405]: Do you think we can add a check for it to the pre-commit hook?

[the remark this answers is not in the document]
Comment (maheichyk) [reply yarn.lock:405]: Maybe we can at some point.

File: src/components/BoardBar/BoardBar.test.tsx [hunk 6]
```diff
@@ -203,6 +203,7 @@ describe('<BoardBar/>', () => {
       version: 'net.nordeck.whiteboard@v1',
       whiteboard: {
         slides: [{ elements: [] }, { elements: [] }, { elements: [] }],
+        files: [],
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply src/components/BoardBar/BoardBar.test.tsx:206]: The `files` prop is now skipped if there are no files.

File: src/components/BoardBar/ExportWhiteboardDialogDownloadFile.tsx (hunk 2 above)
Comment (weeman1337) [reply src/components/BoardBar/ExportWhiteboardDialogDownloadFile.tsx:54]: If a download is not possible at all, we should eventually change the entire export modal to something like „Something is wrong, you cannot export your board“.

I will add it to the ticket for error handling. Would that be okay for you?

File: src/lib/download.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Convert data to JSON and trigger a download.
+ *
+ * @param filename - Filename of the download
+ * @param data - Data to download
+ *
+ * @returns void
+ */
+export function downloadData(filename: string, data: object): void {
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply src/lib/download.ts:25]: Moved :heavy_check_mark:

File: yarn.lock (hunk 5 above)
[the remark this answers is not in the document]
Comment (weeman1337) [reply yarn.lock:405]: I've applied `yarn deduplicate`

File: src/components/BoardBar/ExportWhiteboardDialogDownloadFile.tsx (hunk 2 above)
Comment (weeman1337) [reply src/components/BoardBar/ExportWhiteboardDialogDownloadFile.tsx]: It does now raise an error.

**Review by maheichyk — APPROVED** [review]

**Pull request merged**